src/compute-client/src/service.rs

```diff
@@ -211,17 +210,28 @@ where
                 let mut new_uppers = Vec::new();
 
                 for (id, new_shard_upper) in list {
-                    if let Some((frontier, shard_frontiers)) = self.uppers.get_mut(&id) {
-                        let old_upper = frontier.frontier().to_owned();
-                        let shard_upper = &mut shard_frontiers[shard_id];
-                        frontier.update_iter(shard_upper.iter().map(|t| (t.clone(), -1)));
-                        frontier.update_iter(new_shard_upper.iter().map(|t| (t.clone(), 1)));
-                        shard_upper.join_assign(&new_shard_upper);
-
-                        let new_upper = frontier.frontier();
-                        if PartialOrder::less_than(&old_upper.borrow(), &new_upper) {
-                            new_uppers.push((id, new_upper.to_owned()));
-                        }
+                    // Initialize frontier tracking state for this collection, if necessary.
+                    if !self.uppers.contains_key(&id) {
+                        self.start_frontier_tracking(id);
+                    }
+
+                    let (frontier, shard_frontiers) = self.uppers.get_mut(&id).unwrap();
+
+                    let old_upper = frontier.frontier().to_owned();
+                    let shard_upper = &mut shard_frontiers[shard_id];
+                    frontier.update_iter(shard_upper.iter().map(|t| (t.clone(), -1)));
+                    frontier.update_iter(new_shard_upper.iter().map(|t| (t.clone(), 1)));
+                    shard_upper.join_assign(&new_shard_upper);
+
+                    let new_upper = frontier.frontier();
+                    if PartialOrder::less_than(&old_upper.borrow(), &new_upper) {
+                        new_uppers.push((id, new_upper.to_owned()));
+                    }
+
+                    if new_upper.is_empty() {
+                        // All shards have reported advancement to the empty frontier, so we do not
+                        // expect further updates for this collection.
+                        self.cease_frontier_tracking(id);
                     }
```

Comment:
For the cleanup to work for a given collection, the following must be true:
* All shards eventually report that the collection has advanced to the empty frontier.
* After a shard reported the empty frontier, it emits no more `FrontierUppers` for the collection.

I _think_ both of these assumptions hold, but I find it hard to be sure.